pow/randomx/src/lib.rs

```diff
@@ -99,32 +104,49 @@ unsafe impl<M: WithCacheMode> Send for Cache<M> { }
 unsafe impl<M: WithCacheMode> Sync for Cache<M> { }
 
 impl<M: WithCacheMode> Cache<M> {
-	pub fn new(key: &[u8], config: &Config) -> Self {
+	pub fn new(config: &Config) -> Result<Self, Error> {
 		let flags = M::randomx_flags(config);
 
-		let cache_ptr = unsafe {
-			let ptr = sys::randomx_alloc_cache(flags);
+		let (cache_ptr, dataset_ptr) = unsafe {
+			let dataset_ptr = if M::has_dataset() {
+				let ptr = sys::randomx_alloc_dataset(flags);
+				if ptr.is_null() {
+					return Err(Error::CacheAllocFailed)
+				}
+				Some(ptr)
+			} else {
+				None
+			};
+
+			let cache_ptr = sys::randomx_alloc_cache(flags);
+			if cache_ptr.is_null() {
+				if M::has_dataset() {
+					sys::randomx_release_dataset(dataset_ptr.unwrap());
```

Comment:
Please remove this `unwrap`!

On top `data_ptr` is guaranteed to be `Some`. So you can remove the `Option` type there.

Reply:
And why do we `randomx_alloc_dataset` and then immediately `randomx_release_dataset` in this function? I don't think the logic here is correct.

Reply:
Here, if `sys::randomx_alloc_cache` failed and `sys::randomx_alloc_dataset` was already succesfull then it's releasing the dataset before returing `Err(Error::CacheAllocFailed)`

Reply:
> 
> 
> Please remove this `unwrap`!
> 
> ~On top `data_ptr` is guaranteed to be `Some`. So you can remove the `Option` type there.~
> 
> I'd suggest separate the code if `M::has_dataset` is true or false. If that doesn't work, you can also change this line to be an expect.

👌